clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepoImpl.kt

```diff
@@ -79,23 +71,33 @@ internal class FileResourcesRepoImpl(
 
         synchronized(fetchAllFilesLock) {
             if (downloadInProgressUrls.isEmpty()) {
-                urls.forEach { url ->
+                urlMeta.forEach { (url,_) ->
                     downloadInProgressUrls[url] = mutableListOf(newCompletionCallback)
                 }
-                preloaderStrategy.preloadFiles(urls, successBlock, failureBlock)
+                preloaderStrategy.preloadFilesAndCache(
+                    urlMetas = urlMeta,
+                    successBlock = successBlock,
+                    failureBlock = failureBlock
+                )
             } else {
-                val filteredUrls = mutableListOf<String>()
-                urls.forEach { url ->
+                val filteredUrlsMeta = mutableListOf<Pair<String,CtCacheType>>()
+                urlMeta.forEach { meta ->
+                    val url = meta.first
                     if (downloadInProgressUrls.containsKey(url)) {
                         downloadInProgressUrls[url]?.add(newCompletionCallback)
                     } else {
                         downloadInProgressUrls[url] = mutableListOf(newCompletionCallback)
-                        filteredUrls.add(url)
+                        filteredUrlsMeta.add(meta)
                     }
                 }
-                preloaderStrategy.preloadFiles(filteredUrls, successBlock, failureBlock)
+                preloaderStrategy.preloadFilesAndCache(
+                    urlMetas = filteredUrlsMeta,
+                    successBlock = successBlock,
+                    failureBlock = failureBlock
+                )
             }
```

Comment:
To listen to full completion of all tasks we need not do locking and syncronisation ourselves. This implementation can be used out of the box by using apis of coroutines and executors instead of manually implementing.

1. for coroutine you can launch async and awaitAll to wait for completion and invoke a final callback
2. For executors you can use invokeAlll() method as well. ref https://stackoverflow.com/questions/3269445/executorservice-how-to-wait-for-all-tasks-to-finish